### DISCUSSION TITLE
Password shucking vulnerability

### DISCUSSION THREAD

**Discussion by michaelcbrook** [meteor/meteor#11812] [Feature Requests]
Hi there,

Recently, we underwent a third-party penetration and vulnerability assessment for one of our apps made with Meteor. One of the vulnerabilities that was found had to do with the fact that when users log into the Meteor app with a password, Meteor first hashes the password on the client side with SHA-256 and then hashes that result again with bcrypt.

The tester's feedback was that this practice makes us vulnerable to [password shucking](https://www.scottbrady91.com/authentication/beware-of-password-shucking).

Does anyone have any thoughts on how this could be resolved long term for all Meteor apps?

### References

https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pre-hashing-passwords
https://blog.ircmaxell.com/2015/03/security-issue-combining-bcrypt-with.html
https://www.scottbrady91.com/authentication/beware-of-password-shucking
https://security.stackexchange.com/a/184090

**Comment by WilliamKelley** [comment]
Very interesting attack. I am interested in the response to this as well.

I'm not a security-expert at all, but I can think of two strategies to solve this vuln.

1. stop pre-hashing ~~on client~~
    - Completely avoids password shucking
    - But, passwords must be sent in plain-text client-to-server.
      - That should be secure given TLS
      - I'm guessing pre-hashing was first implemented by meteor/accounts-password to account for contexts where client-to-server communication is unencrypted. before SSL/TLS was de-facto standard. If there's other reasons, to do so, I'm unaware.
      - Even with TLS, plain-text transmission means they can be plainly inspected by server admins
2. ~~pre-hash with a salt on client~~
    - ~~Presumes a viable strategy for secure salting on the client -- not sure what exactly that would be~~
    - ~~A user's existing hash from another site would not much their salted pre-hash as our client -- avoids a necessary condition for shucking~~
    - ~~(note: im much less confident in this being an actual solution or viable)~~

Then there's figuring strategy for migrating existing passwords hashes in the DB. Not sure about that.

_Edit:_ struckthrough "on client" to clarify original intent that pre-hashing shouldn't be done at all.

_Edit:_ struckthrough option 2 after further reflection and conversation

**Reply by timmc** [reply]
Moving the SHA256 from the client to the server side only mitigates some aspects of the vulnerability, but would be a very quick fix (if transmission of plaintext to server is acceptable).

Salting on the client side would either require a two-step login where the salt is sent from the server after the username is entered, or... just using the username as salt. :-) But any method of salting that initial hash necessarily involves a migration, and if there's to be a migration, you might as well take an entirely different approach instead.

I mentioned the idea of using an [aPAKE](https://en.wikipedia.org/wiki/Password-authenticated_key_agreement)s elsethread, since essentially all the hashing is done on the client, and the credential is never transmitted to the server, hashed or otherwise. (The client instead proves that it _possesses_ the credential.) In particular [OPAQUE](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-opaque-02) might be worth investigating, as it allows offloading bcrypt-like "hardening" computation to the client to avoid server DoS vulnerability—although this has its limitations, as the client may have relatively less compute than either server or attacker. Anyway, not something I've looked into deeply, but this might solve several needs at once, if all of them _are_ in fact needs. :-)

**Reply by WilliamKelley** [reply]
> Moving the SHA256 from the client to the server side...  

I wasn't clear initially but this is not what i meant. I meant stop pre-hashing in general, not on client, not on server... feed plaintext directly to bcrypt on server.

Exactly how that works with existing hashes or for a migration, i'm not sure.

As for the second method, yeah i dont think its good after reflection either.

**Reply by timmc** [reply]
Understood―just suggesting a possible partial stopgap. (I'm not actually a fan of the prehashing either.)

**Reply by xet7** [reply]
@WilliamKelley

> Exactly how that works with existing hashes or for a migration, i'm not sure.

I have example of migration etc here:
https://github.com/meteor/meteor/discussions/11812#discussioncomment-1855829

**Comment by xet7** [comment]
But is there any way for that to be remote vulnerability?

I think logging in with Meteor works this way:
1. Login at login page
2. Serverside check that password is correct
3. User is logged in
4. If user tries brute force login, then have lockout after some attempts. I did [remove extra files from atmospherejs package lucasantoniassi:accounts-lockout](https://github.com/wekan/wekan/commit/23e5e1e3bd081699ce39ce5887db7e612616014d) so that it's only [these 10 files that are required for lockout](https://github.com/wekan/wekan/tree/master/packages/wekan-accounts-lockout) and it [reads settings of how many tries allowed as environment variables from source/snap/docker here](https://github.com/wekan/wekan/blob/master/server/accounts-lockout.js).

Is this only if attacker has access to serverside filesystem? But if that Meteor app is running in inside container inside VM that uses full disk encryption, what way would attacker view those passwords?

Or can attacker somehow do that attack remotely clientside? If there is no publication that publishes all usernames/passwords to browserside, how would that attack then work?

**Reply by WilliamKelley** [reply]
I think there is a remote vuln here, but account-specific not exactly DB-wide at once.

Prerequisites:

- meteor's hashing in summary is `bcrypt(sha256(password))` where SHA256 is done client-side and bcrypt is done server-side
- attacker has access to other DB leaks: emails/usernames & password hashes
  - those password hashes are SHA256
- an account for your service has a username/email & underlying password that matches a leaked email/username & hash (of that underlying password)

Attack:

Attacker bypasses your client-side UI / scripts, sending the username/email and SHA256 password hash directly to the server in the data structure that the server expects. Attacker successfully logs-in, without knowing the user's password, only having the hash. 

Furthermore, the attacker can then focus their computational efforts on cracking the SHA256 hash to get the user's plaintext password.

---

But maybe that's not a vulnerability because the user shouldn't be reusing the password?

_Edit: yeah i think this is just a standard vuln, not really shucking._

**Reply by xet7** [reply]
@WilliamKelley 

If client does too much with SHA256, should client only send plaintext password?

If client does too little, should there browserside public key crypto with client generated keys, and all data transfer between client and server always with public key crypto encrypted blobs even with HTTP only?

Of if server does too little, should it be `Argon2(bcrypt(sha256(password)))` ?

What is current best practice for this?

**Reply by xet7** [reply]
Or alternatively, `Argon2(per-meteor-install-env-setting-salt + bcrypt(sha256(password)))` ?

**Reply by xet7** [reply]
If someone uses any up-to-date online password manager, those usually do warn does password exist in any leaked passwords databases.

**Reply by WilliamKelley** [reply]
> If client does too much with SHA256, should client only send plaintext password?

Yes I think it should only send plaintext.

I don't think it's a matter of the client or server doing too much or too little. My scenario (and more largely shucking) only really requires that more than one hashing algo be used on a password (that's a simplification for shucking, but I hope you see the point). I think the sequence and combination is irrelevant, _other than the assumption_ that the inner hash is weaker than the outer hash.

Any modern app should connect exclusively over SSL/TLS and thus the password can be considered secure over transmission even in plaintext.

> What is current best practice for this?

I don't know, I'm not a security expect unfortunately. But I've read discussion around this vulnerability from other sources, and people have said that services as big as Google send passwords in plaintext from the client (secure-transmission of course), and that precludes them from this scenario. So plaintext transmission is my best guess at the current best-practice.

**Reply by xet7** [reply]
Hmm, I did find article that seems to explain how to do hashing properly. I don't know is it a good idea to post about it here, but I did not find any better source. https://crackstation.net/hashing-security.htm

**Reply by WilliamKelley** [reply]
I don't think discussing hashing itself is relevant. This is just to do with the simple practice of hashing hashes. Bcrypt is suitably secure -- you can find that fact in [the original presentation on shucking](https://www.youtube.com/watch?v=OQD3qDYMyYQ) itself.

**Reply by xet7** [reply]
After reading about how to do this properly, would this pseudocode work well enough?

Clientside:

- Send password as plaintext to server

Serverside:
```
// new database fields
models/users.js: add new field/column `salt`
models/settings.js: add field for one string of `per-server-salt`

// If server-salt exist at environment variable, use it
if (env-server-salt !== '') {
  per-server-salt = env-server-salt;

// If per-server-salt exists at database, use it
} else if (settings-server-salt != '') {
  per-server-salt = settings-server-salt;

// If per-server-salt does not exist in database, create it and use it
} else if (settings-server-salt === '') {
   settings-server-salt = random();
   // Generate always new salt. If salt exists at database, generate new until it does not exist at database.
   while (settings-server-salt !== settings-server-salt-already-exists-at-database) {
     settings-server-salt = random();
   }    
   save(settings-server-salt);
   per-server-salt = settings-server-salt;
}
   
// If password is stored to database using old way, and salt does not exist at database
if  (password === bcrypt(sha256(password-at-database)) && per-user-salt-at-database === '') {

   // Generate always new salt. If salt exists at database, generate new until it does not exist at database.
   per-user-salt = random();
   while (per-user-salt !== per-user-salt-already-exists-at-database) {
     per-user-salt = random();
   }    
   per-user-saltedhashedpassword = Argon2(per-server-salt + per-user-salt + password);
   save(per-user-saltedhashedpassword, per-user-salt);
   login-allow();

// If password is stored in new way, allow login
} else if (password === Argon2(per-server-salt + per-user-saltedhashedpassword-at-database + per-user-salt-at-database)) {
   login-allow();

// If no valid password, deny login
} else {
   login-deny();
}
```

**Reply by WilliamKelley** [reply]
I don't understand why you're introducing extraneous salting. Bcrypt does salting internally. It's a secure algo by itself.

I would assume it's a matter of checking:

if stored `hash` equals `bcrypt(password)`? login. 

else equals `bcrypt(sha256(password))`? replace stored hash and login.

else fail.

Then you're left with code coveing both cases for time-eternal, and a shucking vuln for "legacy" hashes.

**Reply by xet7** [reply]
@WilliamKelley 

It seems that unsalted hashes are already available to search at https://crackstation.net . That extra salting is based on advice from https://crackstation.net/hashing-security.htm .

**Reply by WilliamKelley** [reply]
I think you're still missing this: bcrypt salts. The salt and other meta info is stored in the initial bytes of the hash. Introducing another salt is non-standard, unnecessary complexity, in my opinion. I don't think discussing on hashing internals itself is warranted here. The shucking vulnerability and my scenario vuln lies above that.

**Reply by xet7** [reply]
@WilliamKelley 

From my understanding, vulnerability is because hashing is done clientside with hashes that are without salt.

To fix it, clientside passwords are not hashed at all, and at serverside there is per-server-salt and per-client-salt that makes these kinds of brute force attacks unpractical.

Is there any other way to fix this?

**Reply by WilliamKelley** [reply]
> ... vulnerability is because hashing is done clientside with hashes that are without salt.


In my scenario, yes I'd say that's true. For shucking in general, I'd say not really, it's because that >1 hashing algo is chained.

With that difference, eliminating chained hashing in favor of a single, secure, server-side hashing will solve both my scenario and shucking. I think that is the solution.

> To fix it ... that makes these kinds of brute force attacks unpractical.

I don't really want to judge whether or not something is practical. I think that's better left to security experts. A solution like I suggested that avoid introducing extraneous measures is what I understand to be the advised "fix" for shucking (and would fix my scenario.

**Reply by xet7** [reply]
@WilliamKelley

> I don't understand why you're introducing extraneous salting. Bcrypt does salting internally. It's a secure algo by itself.
> 
> I would assume it's a matter of checking:
> 
> if stored `hash` equals `bcrypt(password)`? login.
> 
> else equals `bcrypt(sha256(password))`? replace stored hash and login.
> 
> else fail.
> 
> Then you're left with code coveing both cases for time-eternal, and a shucking vuln for "legacy" hashes.

Ok I did re-read that, and it looks like brute-forcing bcrypt is not practical. So my added extra salting is unnecessary. Thanks!

**Reply by xet7** [reply]
Oops, sorry about my use of word practical. I need to find some other way to write that correctly.

**Comment by xet7** [comment]
Also, Meteor is compatible with Percona MongoDB that has data-at-rest encryption https://www.percona.com/software/mongodb/feature-comparison . Would that help?

**Comment by xet7** [comment]
Or is this about DoS ? 

https://blog.ircmaxell.com/2015/03/security-issue-combining-bcrypt-with.html

"This can lead to a DoS attack where an attacker can provide an exceedingly long password and tie up computer resources."

**Comment by michaelcbrook** [comment]
This is the feedback I received from the tester:

> One option is to reach out for a feature request to Meteor and see if there is a way to change the password storage process and see if there is a way to update the algorithm to one with high work factors and higher entropy values (such as Argon2)

Without that, his immediate recommendation was to audit our database security, as that would be the next best thing. That gets us by for right now and will satisfy the tester, but it doesn't solve the underlying problem.

@xet7 This would mostly be a problem if the database was leaked. And not specifically about DoS. If I understand password shucking correctly, it's where, by an attacker cracking one password, it tells them all they need to know in order to bypass the outer hashing algorithm completely, and then they are able to crack all passwords across accounts more quickly because only the inner hash needs to be cracked and the inner hash (SHA-256) can be calculated at much faster speed than bcrypt, especially if it's unsalted.

I've researched this quite a bit and it's difficult finding the single "best" solution to this because each solution comes with caveats. But the one thing I haven't seen anyone bring up before is the idea of using alternative hashing algorithms, like Argon2. So I wonder if that is new information that would fit somewhere in this topic.

**Reply by xet7** [reply]
@michaelcbrook

Thanks! So would this be only problem with accounts-password, because other auth methods save passwords elsewhere, like LDAP etc?

For accounts-password, changes could maybe be made somewhere here?

- Client: https://github.com/meteor/meteor/blob/devel/packages/accounts-password/password_client.js#L57
- Server: https://github.com/meteor/meteor/blob/devel/packages/accounts-password/password_server.js#L35

**Reply by xet7** [reply]
Although, if Meteor hashes passwords before saving to LDAP etc, then I think this would still affect LDAP etc.

**Reply by WilliamKelley** [reply]
From my understanding, shucking is possible whenever more than one hashing algo is run on a password. So yes definitely accounts-password since it does SHA256 and bcrypt. And if that condition is true for any other accounts service, then shucking is also a problem for that system, but that may or may not be Meteor's problem if it's involved with an external service.

**Reply by timmc** [reply]
I've read through the scottbrady91 post, and I think the claimed scenario is this:

1. Your password table has already been breached by an attacker (as a condition for this attack to work).
2. Some sizeable fraction of other leaked password DBs available to an attacker use the weak, unsalted pre-hash (here, SHA256) rather than plaintext (or a strong hash).
3. The attacker will now go through each bcrypt hash in your DB and check every SHA256 they've seen elsewhere against it. If there's a match, they then try to crack the SHA256 by brute-forcing with dictionary attacks etc. Repeat for the next bcrypt hash, which has a different salt and therefore takes just as long (on average).
4. This benefits the attacker because in addition to leaked cleartext passwords, they can also match up some leaked hashes, and try cracking those (rather than just trying to crack all leaked hashes ahead of time).

Stating a different way, the vulnerability is that if your password hashing *starts with* some weak hashing on the way, leaks from sites that use that identical weak hashing increase the pool of available weak passwords by some margin. Cracking bcrypt(SHA256(A)) is easy if you have SHA256(A) *or* A.

Does that sound like a correct characterization? Have I captured everything? If so, I don't think this is really a vulnerability for Meteor or similar systems, but I want to make sure I'm not barking up the wrong tree first.

(I'd like to *separately* discuss live-system attacks, where client-side pre-hashing might give the attacker some additional reach.)

**Reply by WilliamKelley** [reply]
> If so, I don't think this is really a vulnerability for Meteor or similar systems

I disagree. Just the fact that Meteor Accounts Password's hash algo summary is `bcrypt(sha256(password))` means shucking is a vulnerability. 

I think you've correctly characterized the attack, but nothings negates the fact that the service's algo summary is vulnerable and could be hardened if the summary was just `bcrypt(password)`.

P.S. check out the comment thread with @xet7 and myself above for the separate live-system attack. maybe we can discuss that there if it matches what you're thinking.

**Reply by timmc** [reply]
OK, here's why I don't really see this as being a serious weakness for Meteor's client-side pre-hashing:

The attacker would have to run bcrypt millions of times for each bcrypt hash they find in your DB. I don't know how many SHA256 hashes are available to our hypothetical attacker, but the Pwned Passwords dataset is freely available and contains 630 million SHA1 hashes. If you're using standard hardware and a bcrypt work factor of 10 (which appears to be the default in Meteor), each attempt might take about 100 ms. That's 2 years of compute time just to check a single user's bcrypt hash, with no guarantee of success. And even if they find a match, there's still no guarantee they can crack the SHA256 hash. That will take even more time to attempt. The attacker will parallelize this and use good hardware, but this is already adding up quickly.

Note that that 2 years only includes the time for checking SHA256. They'll also want to spend additional time running bcrypt on leaked cleartext passwords (or SHA256 of those, just to deduplicate).

None of their work for that one user will help them with the next one, excepting the scenario where they match and crack a SHA256. In that case, they can just add it to their plaintext dictionary.

On top of that, there are already people busy at work cracking those leaked SHA256 hashes, [EDIT: some of whom publish] what they find. https://hashes.com and https://crackstation.net/ offer hash reversing based on lookup tables; CrackStation boasts 1-10 billion precomputed hashes, and says that they were [able to crack about half a customer's hashes](https://crackstation.net/crackstation-wordlist-password-cracking-dictionary.htm). [EDIT to add: CrackStation also offers their main 1.5 billion word list for download.]  An attacker is most likely going to be able to rely on the leaked hashes already being cracked, and any SHA256 hashes the attacker can't find in one of those lookup tables is probably one that they're going to have great difficulty cracking themselves (i.e. a hash of a strong password).

Because of all this, I think an attacker would instead spend all their time running known *cleartext* passwords through bcrypt, prioritized by observed frequency in leaks. (Only if that fails would they bother with password shucking.) But bcrypt is already supposed to be strong against dictionary attacks, so if those compute-seconds don't add up high enough, the work factor should be higher!

Given that in reality people choose pretty crackable passwords, avoiding this attack (in a pre-hashed passwords scenario) gives marginal benefit, and I think that's outweighed by the argument to pre-hash on the client side if you don't know if TLS will be in use in real-world deployments.

(Really, I'd rather see Meteor use a PAKE so that you get the best of both worlds, but that's a different topic...)

**Reply by timmc** [reply]
(Added two substantive edits in hash crackers paragraph, marked with "EDIT".)

**Reply by WilliamKelley** [reply]
Sure, whether or not an application wants to address it is a matter of opinion. But the fact that it's a security vulnerability, i.e. weaker than just encrypting with bcrypt straight out, is still true.

The discussion was created based on feedback from a security assessment, and they advised addressing it. From my research, the broader security community considers this a real risk to include in assessment.

I would like to see the risk mitigated, as I'm sure OP would as well. I hope Meteor and the larger community will make every effort to mitigate these risks.

**Reply by timmc** [reply]
I'd agree that it technically creates a weakening, but it's important to weigh it against whatever considerations lead Meteor to choose client-side pre-hashing in the first place. If those considerations are no longer true, or can be addressed in some other way, I would then agree that getting rid of pre-hashing would be good. (Both to remove this epsilon of weakness, and to reduce complexity.)

**Reply by timmc** [reply]
Hmm! I should also temper one of my claims: The attacker doesn't actually need to crack the SHA256 itself if all they want to do is log into the same Meteor instance in the future -- because the SHA256 is done client-side. (This is related to the live-system attack discussed in the other thread, which is indeed what I had in mind.) This means that the attacker *would* be able to use "recalcitrant" hashes that no one had been able to hash, but of course only on that one Meteor site.

I'm not sure what kind of increased pool of credentials this gives the attacker. 20% more? 100%? 200%?

Once the attack is discovered and all passwords reset, this advantage disappears, except for users who *reuse their old password*. So you'd definitely want to protect against reuse across the breach boundary.

**Reply by timmc** [reply]
(I think the implication here is that if Meteor simply switched the SHA256 part from client-side to server-side, the attacker would then be back to having to crack both bcrypt and SHA256. Does that sound right?)

**Reply by WilliamKelley** [reply]
> ... whatever considerations lead Meteor to choose client-side pre-hashing in the first place. 

I am curious about this as well. I hypothesized in my original comment to this broader post that it was done because SSL/TLS wasn't industry standard yet, and thus the transport layer is more vulnerable. But I would like official word on that.

Nonetheless, it is still the case the major services and industry security standards involve transmitting the password in plaintext from client to server. From that we can assume there isn't a meaningful consideration on modern conditions. But we should confirm exactly what modern conditions should be satisfied since Meteor is a suitably old platform.

**Reply by timmc** [reply]
For reference, additional info on bcrypt guessing:

https://openwall.info/wiki/john/benchmarks has some benchmarks for bcrypt using cost factor 5. Using an AWS `c5a.24xlarge` (48 physical cores) appears to allow 86832 guesses per second. For bcrypt cost factor 10 that would be `86832 / 2^(10-5)` or 2713 hash/s. At current AWS spot pricing for that instance type ($1.73/hr) it would take $89 and a bit over 2 days to check 500 million SHA256 hashes against bcrypt for one user.

(The same logic applies to cleartext passwords; even with bcrypt, credential stuffing is pretty damn effective!)

**Reply by timmc** [reply]
If the attacker finds a matching SHA256 and then needs to crack it (say, in the scenario where Meteor moves the SHA256 client-side) then Raw-SHA256 from https://www.openwall.com/john/cloud/c5a.24xlarge-speeds.txt says that same hardware could check 37.4 million (2^25) passwords per second. If these are hard passwords (10 chars of mixed alphanumeric, or ~6 words from a 13K word dictionary) that would take 2^60 tries, or 1100 years. For only moderately strong passwords, say three random words from a 13K word dictionary, we would be looking at 2^41 tries, or 18 hours. (Or for only 2 words: 4 seconds. 🙃) But I don't know where the "pre-cracked" dividing line would be. Anyway, this indicates that the SHA256 cracking part would only be comparable to the bcrypt-guessing step if the recalcitrant passwords were mostly in the 3-4 word range.

**Reply by WilliamKelley** [reply]
I don't really want to down and dirty with calculating work factors of password cracking; it seems beside-the-fact that shucking is a correctly assessed vuln of the system. However, I will say that, on close review, an assumption that you're running with seems wrong to me:

> 3. The attacker will now go through each bcrypt hash in your DB and check every SHA256 they've seen elsewhere against it. ...

From the [original talk](https://www.youtube.com/watch?v=OQD3qDYMyYQ) and other descriptions, like [this](https://superuser.com/questions/1561434/how-do-i-crack-a-double-encrypted-hash), running through every known SHA256 hash is not the presumed attack. Instead, given a "source" (other leaks) and "target" (your leak) data, an attacker "structures" or "correlates" those information pieces by a "user" (username/email) and runs only those through bcrypt -- taking the iterations down to several at max.

Another benefit for the attacker is that there exist a significant amount of uncracked, "weak algo" hashes (MD5, SHA1, SHA256, ...). For example, the above mentioned talk cited that ~12% of public md5 hashes were uncracked. Given a leak of your database, if you use layered hashing, then we can say something like 12% of your users are more vulnerable than they would have been if you had not used layered hashing. That's assessed as significant. I assume there's a similar, if not larger, portion of SHA256 hashes that are also uncracked.

**Reply by timmc** [reply]
Ah! OK, that makes a lot more sense now. I didn't watch the video (40 minutes of continuous attention is a big time commitment -- is there a text version somewhere?) which is why I wasn't sure I understood correctly. So really it's running bcrypt a number of times at a low multiple of the number of users—purely assuming users have reused passwords across sites, and not assuming multiple users are using the same (weak) password.

Great to get an estimate on the uncracked hashes.

**Reply by WilliamKelley** [reply]
Unfortunately I haven't found a transcript or written source of equal quality. But the author and company did collab with OWASP and others for their summaries and recommendations. I will say that the talk is not super complex (so doesnt require _full_ attention except one or two slides), and you could get away with skipping a good portion of the intro.

And yes I'd say that informal summary is accurate. 

Another statistic from the talk was that around 70% of users exposed in new breaches are reusing passwords exposed in old breaches. That helps emphasize how central protecting password reuse is for your security. Even though it's bad, it's reality :/

**Comment by harryadel** [comment]
@michaelcbrook You definitely have a solid case that needs attention and I really appreciate you sharing insights not only about your own learning/research but also what the tester advised yet the best way to communicate this is directly to the Meteor team as recommended in their policy [here](https://github.com/meteor/meteor/security/policy). Sharing such information out in public might lead malicious users to use it one way or another.

**Reply by michaelcbrook** [reply]
@harryadel Seeing as though this topic has already been discussed [here](https://github.com/meteor/meteor-feature-requests/issues/44) and other major tech companies like Dropbox also [disclose their own use](https://dropbox.tech/security/how-dropbox-securely-stores-your-passwords) of the same password hashing strategy, I assumed this issue was already known but there was just no great solution as of yet.

I also wouldn't consider this a security "bug". Meteor's password hashing works as intended to the best of the technology of the time. With more research in this area and more hashing algorithms now available, there may be an opportunity to improve the security of passwords in Meteor, but I wouldn't consider this a bug.

Just to be clear, the way Meteor hashes passwords is very common. It is no less secure than most frameworks or software out there, but I believe, like any feature, there's possible room for improvement. And I think the community can help make suggestions to arrive at a great solution.

**Reply by harryadel** [reply]
I see, sorry for misinterpreting things. :)

**Comment by xet7** [comment]
Hi,
about Argon2. Today when I was in progress of translating Gitea from English to Finnish, I noticed this translation string, related to Argon2:

> Set the password hashing algorithm. Algorithms have differing requirements and strength. `argon2` whilst having good characteristics uses a lot of memory and may be inappropriate for small systems.

**Reply by StorytellerCZ** [reply]
Update on Argon2: https://github.com/meteor/meteor/discussions/13529#discussioncomment-11869775

**Comment by paulincai** [comment]
I red through this conversation very interested for 10 minutes just to remember that I use Passwordless :)

**Reply by StorytellerCZ** [reply]
That approach has its own issues as well. Personally I despise it as it slows me down.

**Comment by jankapunkt** [comment]
Coming back to this. Nowadays the Web Crypto API is basically standard for all major browsers. With this we can now safely generate crypto-level random and thus use salting on the client. @StorytellerCZ should we consider this in accordance to #13529 ?

**Reply by StorytellerCZ** [reply]
I believe that is the next logical step.
